File: src/Transforming/AmbiguousDescription.cs
using System.Text;
using System.Text.RegularExpressions;
using LannaUtils.Unicodes;

namespace LannaUtils.Transforming;

public static class AmbiguousDescription
{
    private readonly static CompoundVowelPattern[] _compoundVowelPatterns = CompoundVowelPattern.GetList("compound");

    public readonly static Regex InitialRegex = new Regex(
        $"^((?<special>" + string.Join("|", AmbiguousRepresentation.SpecialValuesDict.Select(x => x.Key)) + "$)?|"
        + $"((?<haCombination>{Patterns.LetterHighHa}{Patterns.Choeng}(?<afterHaLetter>[{Patterns.AfterHaLetters}]))|(?<compoundInitial>(?<compoundMain>[{Patterns.SpatialCharacters}])(?<compoundChoeng>{Patterns.Choeng}(?<compoundChoengLetter>[{Patterns.ChoengableLettersExceptYaWa}]))+)|(?<singleInitial>[{Patterns.SpatialCharacters}]))"
        + $"(?<leftComponent>[{Patterns.LeftComponents}])?(?<lowerComponent>[{Patterns.LowerComponents}{Patterns.LowerToRightComponents}])?(?<maisam>{Patterns.UpperMaisam})?(?<remaining>.*))$"
    );

    public readonly static Regex VowelRegex = new Regex(
        $"^(?<startTone>[{Patterns.Tones}])?(?<vowel>(?<compound>" + (string.Join("|", _compoundVowelPatterns.Select(x => x.Pattern))) + ")"
        + $"|(?<single>(?<singleToneBefore>[{Patterns.Tones}])?(?<singleVowel>[{Patterns.Vowels}])(?<singleToneAfter>[{Patterns.Tones}])?))?"
        + $"(?<remaining>.*)$"
    );

    public readonly static Regex FinalRegex = new Regex(
        $"^(?<any1>.+)?(?<choeng>{Patterns.Choeng}(?<choengLetters>[{Patterns.ChoengableLetters}]))+(?<any2>.+)?$"
    );

    public static string Describe(string word) => string.Join("", Segmenting.TextSegment.Split(word).Select(segment => DescribeSegment(segment)));

    private static string DescribeSegment(string segment) => GetInitial(segment, out segment) + GetVowel(segment, out segment) + GetFinal(segment);

    private static string GetInitial(string segment, out string remainingSegment)
    {
        Match match = InitialRegex.Match(segment);
        if (!match.Success)
        {
            remainingSegment = segment;
            return "?";
        }

        if (match.Groups["special"].Success)
        {
            remainingSegment = "";
            return AmbiguousRepresentation.GetSpecialValue(match.Groups["special"].Value);
        }

        StringBuilder stringBuilder = new StringBuilder();

        if (match.Groups["haCombination"].Success)
        {
            stringBuilder.Append(AmbiguousRepresentation.GetValue(match.Groups["afterHaLetter"].Value));
        }
        else if (match.Groups["singleInitial"].Success)
        {
            stringBuilder.Append(AmbiguousRepresentation.GetValue(match.Groups["singleInitial"].Value));
        }
        else if (match.Groups["compoundInitial"].Success)
        {
            stringBuilder.Append(AmbiguousRepresentation.GetValue(match.Groups["compoundMain"].Value));
            foreach (Capture capture in match.Groups["compoundChoengLetter"].Captures)
            {
                stringBuilder.Append(AmbiguousRepresentation.GetValue(capture.Value));
            }
        }

        if (match.Groups["leftComponent"].Success)
        {
            stringBuilder.Append(AmbiguousRepresentation.GetValue(match.Groups["leftComponent"].Value));
        }
        if (match.Groups["lowerComponent"].Success)
        {
            stringBuilder.Append(AmbiguousRepresentation.GetValue(match.Groups["lowerComponent"].Value));
        }

        remainingSegment = match.Groups["remaining"].Success ? match.Groups["remaining"].Value : "";
        return stringBuilder.ToString();
    }

    private static string GetVowel(string processedSegment, out string remainingSegment)
    {
        Match match = VowelRegex.Match(processedSegment);

        remainingSegment = match.Groups["remaining"].Success ? match.Groups["remaining"].Value : "";

        if (match.Groups["compound"].Success)
        {
            foreach (CompoundVowelPattern pattern in _compoundVowelPatterns)
            {
                if (match.Groups[pattern.GroupName].Success)
                {
                    return pattern.AmbiguousValue;
                }
            }
            return "?";
        }
        else if (match.Groups["single"].Success)
        {
            return AmbiguousRepresentation.GetValue(match.Groups["singleVowel"].Value);
        }

        return "";
    }

    private static string GetFinal(string processedSegment)
    {
        Match match = FinalRegex.Match(processedSegment);

        StringBuilder stringBuilder = new StringBuilder();

        if (match.Groups["any1"].Success)
        {
            stringBuilder.Append(AmbiguousRepresentation.GetValue(match.Groups["any1"].Value));
        }

        if (match.Groups["choeng"].Success)
        {
            foreach (Capture capture in match.Groups["choengLetters"].Captures)
            {
                stringBuilder.Append(AmbiguousRepresentation.GetValue(capture.Value));
            }
        }

        if (match.Groups["any2"].Success)
        {
            stringBuilder.Append(AmbiguousRepresentation.GetValue(match.Groups["any2"].Value));
        }

        return stringBuilder.ToString();
    }
}

File: js/src/ambiguous-description.ts
import resources from './resources/ambiguous-description';
import textSegment from './text-segment';

const initialRegexPattern = resources.initialRegex;
const vowelRegexPattern = resources.vowelRegex;
const finalRegexPattern = resources.finalRegex;
const compoundVowelPatterns = resources.compoundVowelPatterns;
const letterValues = resources.letterValues;
const specialValues = resources.specialValues;

const getLetterValue = (letter: string): string => (letterValues as any)[letter] ?? '?';
const getSpecialValue = (letter: string): string => (specialValues as any)[letter] ?? '?';

const getInitial = (segment: string): {result: string, remaining: string} => {
  const regex = new RegExp(initialRegexPattern, 'gu');
  const match = regex.exec(segment);
  if (!match || !match.groups) {
    return {result: '', remaining: segment};
  }

  if (match.groups['special']) {
    return {result: getSpecialValue(match.groups['special']), remaining: ''};
  }

  let result = '';
  if (match.groups['haCombination']) {
    result += getLetterValue(match.groups['afterHaLetter']);
  } else if (match.groups['singleInitial']) {
    result += getLetterValue(match.groups['singleInitial']);
  } else if (match.groups['compoundInitial']) {
    result += getLetterValue(match.groups['compoundMain']);
    let choengMatch: RegExpExecArray|null = match;
    do {
      result += (choengMatch && choengMatch.groups) ? getLetterValue(choengMatch.groups['compoundChoengLetter'] ?? '') : '';
    }
    while (choengMatch = regex.exec(segment));
  }

  if (match.groups['leftComponent']) {
    result += getLetterValue(match.groups['leftComponent']);
  }
  if (match.groups['lowerComponent']) {
    result += getLetterValue(match.groups['lowerComponent']);
  }

  return {result: result, remaining: match.groups['remaining'] ?? ''};
};

const getVowel = (processedSegment: string): {result: string, remaining: string} => {
  const match = new RegExp(vowelRegexPattern, 'gu').exec(processedSegment);

  if (match && match.groups) {
    if (match.groups['compound']) {
      for (const compoundVowelPattern of compoundVowelPatterns) {
        if (match.groups[compoundVowelPattern.GroupName]) {
          return {result: compoundVowelPattern.AmbiguousValue, remaining: match.groups['remaining'] ?? ''}
        }
      }
    } else if (match.groups['single']) {
      return {result: getLetterValue(match.groups['singleVowel']), remaining: match.groups['remaining'] ?? ''};
    }
  }

  return {result: '', remaining: ''};
};

const getFinal = (processedSegment: string): string => {
  const regex = new RegExp(finalRegexPattern, 'gu');
  const match = regex.exec(processedSegment);
  if (!match || !match.groups) {
    return '';
  }

  let result = '';
  if (match.groups['any1']) {
    result += getLetterValue(match.groups['any1']);
  }
  if (match.groups['choeng']) {
    let choengMatch: RegExpExecArray|null = match;
    do {
      result += choengMatch && choengMatch.groups ? getLetterValue(choengMatch.groups['choengLetters'] ?? '') : '';
    } while (choengMatch = regex.exec(processedSegment));
  }
  if (match.groups['any2']) {
    result += getLetterValue(match.groups['any2']);
  }

  return result;
};

const describeSegment = (segment: string): string => {
  const initial = getInitial(segment);
  const vowel = getVowel(initial.remaining);
  const final = getFinal(vowel.remaining);
  return initial.result + vowel.result + final;
};

export default function (text: string): string {
  return textSegment(text).map(describeSegment).join('');
}
